java/src/main/kotlin/io/spine/validation/java/ConstraintViolations.kt

```diff
@@ -0,0 +1,97 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.validation.java
+
+import io.spine.base.FieldPath
+import io.spine.protodata.ast.TypeName
+import io.spine.protodata.ast.qualifiedName
+import io.spine.protodata.java.ClassName
+import io.spine.protodata.java.Expression
+import io.spine.protodata.java.StringLiteral
+import io.spine.protodata.java.mapExpression
+import io.spine.protodata.java.newBuilder
+import io.spine.protodata.java.packToAny
+import io.spine.validate.ConstraintViolation
+import io.spine.validate.TemplateString
+import io.spine.validate.checkPlaceholdersHasValue
+
+/**
+ * Yields an expression that creates a new instance of [ConstraintViolation]
+ * with the given parameters.
+ *
+ * @param errorMessage The error message template string.
+ * @param declaringType The message type being validated.
+ * @param fieldPath The path to the field containing an invalid value.
+ * @param fieldValue The field value violated the constraint, if any.
+ *   For some validation options, the field value is either absent or irrelevant.
+ *   For example, `(required)` option doesn't set it for its [ConstraintViolation]s.
+ */
+public fun constraintViolation(
+    errorMessage: Expression<TemplateString>,
+    declaringType: TypeName,
+    fieldPath: Expression<FieldPath>,
+    fieldValue: Expression<*>? = null
+): Expression<ConstraintViolation> {
+    var builder = ClassName(ConstraintViolation::class).newBuilder()
+        .chainSet("message", errorMessage)
+        .chainSet("type_name", StringLiteral(declaringType.qualifiedName))
+        .chainSet("field_path", fieldPath)
+    fieldValue?.let {
+        builder = builder.chainSet("field_value", fieldValue.packToAny())
+    }
+    return builder.chainBuild()
+}
+
+/**
+ * Yields an expression that creates a new instance of [TemplateString]
+ * with the given parameters.
+ *
+ * @param template The template string that may have one or more placeholders.
+ * @param placeholders The supported placeholders and their values.
+ * @param optionName The name of the option, which declared the provided [placeholders].
+ * @param fieldName The fully qualified name of the field, which passed the provided [template].
+ */
+public fun templateString(
+    template: String,
+    placeholders: Map<ErrorPlaceholder, Expression<String>>,
+    optionName: String,
+    fieldName: String
+): Expression<TemplateString> {
+    checkPlaceholdersHasValue(template, placeholders.mapKeys { it.key.value }) {
+        "The `($optionName)` option doesn't support the following placeholders: `$it`. " +
+                "The supported placeholders: `${placeholders.keys}`. " +
+                "The declared field: `${fieldName}`."
+    }
+    val placeholderEntries = mapExpression(
+        ClassName(String::class), ClassName(String::class),
+        placeholders.mapKeys { StringLiteral(it.key.toString()) }
+    )
+    return ClassName(TemplateString::class).newBuilder()
```

Comment:
Similarly to the class name of `String`, we probably would have multiple places that build a `TemplateString` instance. So an extension function or a companion method may be handy.

I'm no insisting at the moment because the code is going to change. But please keep it in mind for following PRs.